### PR TITLE
Add xcprivacy privacy manifest to iOS framework

### PR DESCRIPTION
**Pull request by jmagman**
Create a `PrivacyInfo.xcprivacy` (this name is required) plist and move it to the top-level of the iOS framework bundle.  `NSPrivacyTracking*` and `NSPrivacyCollectedDataTypes` keys are required, but the values are blank.

Apple [now requires](https://developer.apple.com/documentation/bundleresources/privacy_manifest_files) that third-party frameworks must include this manifest to document usage of particular APIs and [how they are used](https://developer.apple.com/documentation/bundleresources/privacy_manifest_files/describing_use_of_required_reason_api).

> Starting in fall 2023, when you upload a new app or app update to App Store Connect that uses an API (including from third-party SDKs) that requires a reason, you’ll receive a notice if you haven’t provided an approved reason in your app’s privacy manifest. And starting in spring 2024, in order to upload your new app or app update to App Store Connect, you’ll be required to include an approved reason in the app’s privacy manifest which accurately reflects how your app uses the API.

https://developer.apple.com/news/?id=z6fu1dcu

@stuartmorgan [audited](https://github.com/flutter/flutter/issues/131494#issuecomment-1852718759) the engine and third_party:

> * File timestamps:
>   
>   * `C617.1` for app state restoration in `FlutterAppDelegate.mm`.
>   * `0A2A.1` for implementation of the relevant `File` wrappers.
> * System boot time:
>   
>   * `35F9.1` for various event timing and elapsed time calculations.

Note macOS frameworks do not need to declare `NSPrivacyAccessedAPITypes`.
I don't think this will require recipe or conductor codesign changes since this is a file copied as a resource into the framework, just as the modulemap and Info.plist aren't referenced anywhere.

I'm not quite sure how to test this other than letting it build and generate a `Generate Privacy Report` in a Flutter app in Xcode.  There's no where we check that, say, the Info.plist is copied to the right place in Flutter.framework (even in flutter/flutter).  When this rolls into the framework I will add a check to [ios_content_validation_test.dart](https://github.com/flutter/flutter/blob/3da9bc169837d223496439b6d5f6b7e1a82d4318/dev/devicelab/bin/tasks/ios_content_validation_test.dart)

On this PR the `PrivacyInfo.xcprivacy` was written to the expected location in the Flutter.framework:
https://cas-viewer.appspot.com/projects/chromium-swarm/instances/default_instance/blobs/d45cd0809420f08145c7b78ea96cba6e7ea48d8ecfdc8fd2411f82fa65444714/516/tree

Fixes https://github.com/flutter/flutter/issues/131494

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/darwin/ios/framework/PrivacyInfo.xcprivacy [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+		<dict/>
+	</array>
+	<key>NSPrivacyAccessedAPITypes</key>
+	<array>
+		<dict>
+			<key>NSPrivacyAccessedAPIType</key>
+			<string>NSPrivacyAccessedAPICategoryFileTimestamp</string>
+			<key>NSPrivacyAccessedAPITypeReasons</key>
+			<array>
+				<string>0A2A.1</string>
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/PrivacyInfo.xcprivacy:20]: > Declare this reason if your third-party SDK is providing a wrapper function around file timestamp API(s) for the app to use, and you only access the file timestamp APIs when the app calls your wrapper function. This reason may only be declared by third-party SDKs. This reason may not be declared if your third-party SDK was created primarily to wrap required reason API(s). Information accessed for this reason, or any derived information, may not be used for your third-party SDK’s own purposes or sent off-device by your third-party SDK.

File: shell/platform/darwin/ios/framework/PrivacyInfo.xcprivacy [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+				<string>C617.1</string>
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/PrivacyInfo.xcprivacy:21]: For app state restoration in `FlutterAppDelegate.mm`:
> Declare this reason to access the timestamps, size, or other metadata of files inside the app container, app group container, or the app’s CloudKit container.

File: shell/platform/darwin/ios/framework/PrivacyInfo.xcprivacy [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+			</array>
+		</dict>
+		<dict>
+			<key>NSPrivacyAccessedAPIType</key>
+			<string>NSPrivacyAccessedAPICategorySystemBootTime</string>
+			<key>NSPrivacyAccessedAPITypeReasons</key>
+			<array>
+				<string>35F9.1</string>
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/PrivacyInfo.xcprivacy:29]: > Declare this reason to access the system boot time in order to measure the amount of time that has elapsed between events that occurred within the app or to perform calculations to enable timers. Information accessed for this reason, or any derived information, may not be sent off-device. There is an exception for information about the amount of time that has elapsed between events that occurred within the app, which may be sent off-device.

**Comment by jmagman** [comment]
The `PrivacyInfo.xcprivacy` was written to the expected location in the framework:
https://cas-viewer.appspot.com/projects/chromium-swarm/instances/default_instance/blobs/d45cd0809420f08145c7b78ea96cba6e7ea48d8ecfdc8fd2411f82fa65444714/516/tree

File: shell/platform/darwin/ios/framework/PrivacyInfo.xcprivacy [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+<plist version="1.0">
+<dict>
+	<key>NSPrivacyTracking</key>
+	<false/>
+	<key>NSPrivacyTrackingDomains</key>
+	<array/>
+	<key>NSPrivacyCollectedDataTypes</key>
+	<array>
+		<dict/>
+	</array>
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/PrivacyInfo.xcprivacy:12]: False and empty, we are not tracking or collecting.

**Comment by Hixie** [comment]
might be worth putting the comments explaining these (that you wrote on this PR) literally as XML comments in the file itself.

**Comment by Hixie** [comment]
test-exempt: apple does not provide a way to meaningfully test this

File: shell/platform/darwin/ios/framework/PrivacyInfo.xcprivacy
Comment (matanlurey) [inline shell/platform/darwin/ios/framework/PrivacyInfo.xcprivacy:1]: I only wonder if someone of these review comments are better suited as `<!-- comments -->`, particularly ones with confusing semantics (i.e. that an array of an empty dict means "we are not tracking or collecting").

**Comment by jmagman** [comment]
> literally as XML comments in the file itself

I would do this if we controlled how this file is used, but I'm not confident Apple's privacy tooling won't 💩  on an unexpected comment (or key, or value, or...).  Hopefully this PR is documentation enough for the git archeologists.

I will include a better comment when I write the framework test that validates the file is where we expect it to be.

File: shell/platform/darwin/ios/BUILD.gn [hunk 5]
```diff
@@ -373,6 +373,12 @@ copy("copy_framework_module_map") {
   outputs = [ "$_flutter_framework_dir/Modules/module.modulemap" ]
 }
 
+copy("copy_framework_privacy_manifest") {
```
Comment (stuartmorgan) [inline shell/platform/darwin/ios/BUILD.gn:376]: Maybe a comment linking to https://developer.apple.com/documentation/bundleresources/privacy_manifest_files here to provide context on what this is and why we are doing it?

Although developer.apple.com links are notoriously fragile, so 🤷🏻

**Review by stuartmorgan — APPROVED** [review]
LGTM

File: shell/platform/darwin/ios/BUILD.gn (hunk 5 above)
Comment (jmagman) [reply shell/platform/darwin/ios/BUILD.gn:376]: We don't explain Info.plists, or modulemaps, or the values within those files, or any other thing we move around in this gn file.  In 2 years when you Google "PrivacyInfo.xcprivacy" you'll see the most relevant info, and when you `git blame` you'll find this PR with my thorough description.  However adding the link is harmless so I'll do that.

**Pull request merged**